Add documentation for `jcmd` support (#9987)
oracle/graal: docs/reference-manual/native-image/JCmd.md

This support complements existing Native Image monitoring features, including JDK Flight Recorder, heap dumps, and native memory tracking.

## Enabling `jcmd` Support

Support for `jcmd` is disabled by default and must be explicitly enabled at build time.
 
To build a native executable with `jcmd` support, use the `--enable-monitoring=jcmd` option.
```shell
native-image --enable-monitoring=jcmd YourApplication
```

When enabling support for `jcmd`, you may also want to include additional monitoring features, such as JDK Flight Recorder or heap dumps.
Including multiple monitoring features during the Native Image build process unlocks access to more diagnostic commands at runtime. For example:
```shell
native-image --enable-monitoring=jcmd,jfr,heapdump YourApplication
```

You may also want to include the `jvmstat` monitoring feature so your native executable can be discovered and listed with `jcmd -l` or `jcmd` with no arguments provided.
```shell
native-image --enable-monitoring=jcmd,jvmstat YourApplication
```

```shell

> jcmd -l

Author: oh yes, I'll fix that

| JFR.check                | jfr                                             | Checks running JFR recording(s).                                                                    |
| JFR.dump                 | jfr                                             | Copies contents of a JFR recording to file. Either the name or the recording id must be specified.  |
| Thread.dump_to_file      | Always available                                | Dump threads, with stack traces, to a file in plain text or JSON format.                            |
| Thread.print             | Always available                                | Print all threads with stacktraces.                                                                 |
| VM.command_line          | Always available                                | Print the command line used to start this VM instance.                                              |
| VM.native_memory         | nmt                                             | Print native memory usage.                                                                          |

> | VM.system_properties     | Always available                                | Print native memory usage                                                                           |

Author: Oh yes, good catch. I need to change the description for that one.

---
layout: docs
toc_group: debugging-and-diagnostics
link_title: JCmd
permalink: /reference-manual/native-image/debugging-and-diagnostics/JCmd/
---

# Java Diagnostic Command (jcmd) with Native Image

 Native Image has support for Java Diagnostic Command (`jcmd`). Users can now use the same `jcmd` tool they use to interact with Java applications to interact with Native Image executables. Support for `jcmd` compliments many of the other monitoring features that exist in Native Image such as JDK Flight Recorder, heap dumps, and native memory tracking.  

## Enabling `jcmd` Support

Support for `jcmd` is disabled by default and must be explicitly enabled at build time.
 
To build a native executable with `jcmd` support, use the `--enable-monitoring=jcmd` option.
```shell
native-image --enable-monitoring=jcmd YourApplication
```

> 

Reviewer: Please briefly mention `jvmstat` as well (without `jvmstat` support, `jcmd` won't list the native image process if `jcmd` is executed without any arguments).

Author: Good idea! I've added a note about jvmstat

```
Including more monitoring features at Native Image build time will provide access to more diagnostic commands at runtime.

To use `jcmd` at runtime, start your Native Image executable normally and obtain its process ID (PID).  Once you have the PID, you can use it with `jcmd` to connect with the running Native Image executable. For example, to obtain a list of the commands available in a specific Native Image executable you can execute `jcmd <pid> help`.
```shell
jcmd 388454 help

388454:
The following commands are available:
GC.heap_dump
GC.run
JFR.start
JFR.stop
JFR.check
JFR.dump
Thread.dump_to_file
Thread.print
VM.command_line
VM.native_memory
VM.system_properties
VM.uptime
VM.version
help

For more information about a specific command use 'help <command>'.

```

## Supported Diagnostic Commands
The following key-value pairs are supported:

| Name                 | Included with `--enable-monitoring=` | Description                                                                                        |
|----------------------|--------------------------------------|----------------------------------------------------------------------------------------------------|

> | GC.heap_dump         | heapdump                             | Generate a HPROF format dump of the Java heap.                                                     |

Reviewer: `Compiler.dump_code_cache` is missing here (only available with Truffle).

| Thread.print         | Always available                     | Print all threads with stacktraces.                                                                |
| VM.command_line      | Always available                     | Print the command line used to start this VM instance.                                             |
| VM.native_memory     | nmt                                  | Print native memory usage.                                                                         |
| VM.system_properties | Always available                     | Print native memory usage                                                                          |
| VM.uptime            | Always available                     | Print VM uptime.                                                                                   |
| VM.version           | Always available                     | Print JVM version information.                                                                     | 
| help                 | Always available                     | Display help information.                                                                          |

## Performance

> The adding `jcmd` support will have minimal impact on Native Image performance while idle.  However, the performance impact of the `jcmd` feature greatly depends on which diagnostic commands are invoked and their frequency. For example, invoking multiple garbage collections will have a much greater impact than dumping a single native memory tracking report. 

Reviewer 2: ```suggestion
Adding `jcmd` support to Native Image has minimal impact on performance when the application is idle.
However, the performance impact varies significantly depending on the diagnostic commands used and how frequently they are invoked.
For example, triggering multiple garbage collections will have a much greater overhead than dumping a single native memory tracking report.
```

> The adding `jcmd` support will have minimal impact on Native Image performance while idle.  However, the performance impact of the `jcmd` feature greatly depends on which diagnostic commands are invoked and their frequency. For example, invoking multiple garbage collections will have a much greater impact than dumping a single native memory tracking report. 

Reviewer: Please mention that `jcmd help <cmd>` prints the impact of a command.

> 

Reviewer 2: Double empty lines
```suggestion
```

## Limitations

Currently, this feature is not available on Windows.

### Further Reading

> - [Debugging and Diagnostics](DebuggingAndDiagnostics.md)

Reviewer 2: I would also add some JFR guide to this reading suggestion:
```suggestion
- [Build and Run Native Executables with JFR](guides/build-and-run-native-executable-with-jfr.md)
```

Author: Good idea! Applied!
